electrumsv/network_support/esv_client.py

```diff
@@ -0,0 +1,436 @@
+
+    async def _message_fetcher_job(self) -> None:
+        """Idempotent - if spawned twice, the second time will do nothing"""
+        if not self._message_fetcher_is_alive:
+            peer_channel_notification_queue: asyncio.Queue[GeneralNotification] = asyncio.Queue()
+            for i in range(self._FETCH_JOBS_COUNT):
+                asyncio.create_task(
+                    self._fetch_peer_channel_message_job(peer_channel_notification_queue))
+
+            async for notification in self.subscribe_to_general_notifications():
+                peer_channel_notification_queue.put_nowait(notification)
+
+    async def wait_for_merkle_proofs_and_double_spends(self) -> AsyncIterable[TSCMerkleProof]:
+        if not self._message_fetcher_is_alive:
+            asyncio.create_task(self._message_fetcher_job())
+
+        # https://github.com/bitcoin-sv-specs/brfc-merchantapi#callback-notifications
+        while True:
+            # Todo run select query on MAPIBroadcastCallbacks to get libsodium encryption key
+            callback_response = await self._merkle_proofs_queue.get()
+            tsc_merkle_proof: TSCMerkleProof = json.loads(callback_response['callbackPayload'])
+
+            # Todo caller to delete entry in MAPIBroadcastCallbacks when processed
+            yield tsc_merkle_proof
+
+    async def subscribe_to_general_notifications(self) -> AsyncIterable[GeneralNotification]:
+        """Concurrent fetching of peer channel messages is left to the caller in order to keep
+        this class very simple"""
+        ws_base_url = self._replace_http_with_ws(self.base_url)
+        url = ws_base_url + "api/v1/web-socket" + f"?token={self.master_token}"
+        logger.debug(f"URL IS: {url}")
+        try:
+            async with self.session.ws_connect(url, headers={}, timeout=5.0) as ws:
+                logger.info(f'Connected to {url}')
+                msg: aiohttp.WSMessage
+                async for msg in ws:
+                    if msg.type == aiohttp.WSMsgType.TEXT:
+                        notification: GeneralNotification = json.loads(msg.data)
+                        yield notification
+
+                    if msg.type in (aiohttp.WSMsgType.CLOSE, aiohttp.WSMsgType.ERROR,
+                            aiohttp.WSMsgType.CLOSED, aiohttp.WSMsgType.CLOSING):
+                        logger.info("CLOSED")
+                        break
+        except WSServerHandshakeError as e:
+            if e.status == 401:
+                raise WebsocketUnauthorizedException()
+
+    # ----- HeaderSV APIs ----- #
+    async def get_single_header(self, block_hash: bytes) -> bytes:
+        url = self.base_url + f"api/v1/headers/{hash_to_hex_str(block_hash)}"
+        headers = {}
+        headers.update(self.headers)
+        headers.update({"Accept": "application/octet-stream"})
+        async with self.session.get(url, headers=headers) as resp:
+            resp.raise_for_status()
```

Comment:
`raise_for_status` calls give me the willies. What are the repercussions on the callers, needs of the callers?

Reply:
Fair enough but I don't want to think about the fine details of how each and every client request will be handled because I am not yet using them for anything. Remember this is Part 1 of 3 parts...

Reply:
How about a policy that whenever one of these ESVClient methods is actually used somewhere, the `resp.raise_for_status()` must be removed and all outcomes handled properly by the callers?

Reply:
And then the caller will need to handle the None when we get to actually using these requests... it might be more appropriate to raise a custom exception... It's all up in the air and I don't really want to deal with it yet.